crates/luminol-filesystem/src/web.rs

```diff
@@ -43,127 +42,119 @@ pub struct FileSystemCommand(FileSystemCommandInner);
+        flume::Sender<Result<Vec<DirEntry>>>,
     ),
-    DirDrop(usize, oneshot::Sender<bool>),
-    DirClone(usize, oneshot::Sender<usize>),
-    FileRead(usize, usize, oneshot::Sender<std::io::Result<Vec<u8>>>),
-    FileWrite(usize, Vec<u8>, oneshot::Sender<std::io::Result<()>>),
-    FileFlush(usize, oneshot::Sender<std::io::Result<()>>),
+    DirDrop(usize, flume::Sender<bool>),
+    DirClone(usize, flume::Sender<usize>),
+    FileRead(usize, usize, flume::Sender<std::io::Result<Vec<u8>>>),
+    FileWrite(usize, Vec<u8>, flume::Sender<std::io::Result<()>>),
+    FileFlush(usize, flume::Sender<std::io::Result<()>>),
     FileSeek(
         usize,
         std::io::SeekFrom,
-        oneshot::Sender<std::io::Result<u64>>,
+        flume::Sender<std::io::Result<u64>>,
     ),
-    FileDrop(usize, oneshot::Sender<bool>),
+    FileDrop(usize, flume::Sender<bool>),
+    FileSize(usize, flume::Sender<u64>),
 }
 
-fn filesystem_tx_or_die() -> &'static mpsc::UnboundedSender<FileSystemCommand> {
+fn filesystem_tx_or_die() -> &'static flume::Sender<FileSystemCommand> {
     FILESYSTEM_TX.get().expect("FileSystem sender has not been initialized! Please call `FileSystem::initialize_sender` before calling this function.")
 }
 
 impl FileSystem {
     /// Initializes the sender that we use to send filesystem commands to the main thread.
     /// This must be called before performing any filesystem operations.
-    pub fn setup_filesystem_sender(filesystem_tx: mpsc::UnboundedSender<FileSystemCommand>) {
+    pub fn setup_filesystem_sender(filesystem_tx: flume::Sender<FileSystemCommand>) {
         FILESYSTEM_TX
             .set(filesystem_tx)
             .expect("FileSystem sender cannot be initialized twice");
     }
 
     /// Returns whether or not the user's browser supports the JavaScript File System API.
     pub fn filesystem_supported() -> bool {
-        let (oneshot_tx, oneshot_rx) = oneshot::channel();
+        let (oneshot_tx, oneshot_rx) = flume::bounded(1);
         filesystem_tx_or_die()
             .send(FileSystemCommand(FileSystemCommandInner::Supported(
                 oneshot_tx,
             )))
             .unwrap();
-        oneshot_rx.blocking_recv().unwrap()
+        oneshot_rx.recv().unwrap()
     }
 
     /// Attempts to prompt the user to choose a directory from their local machine using the
     /// JavaScript File System API.
     /// Then creates a `FileSystem` allowing read-write access to that directory if they chose one
     /// successfully.
     /// If the File System API is not supported, this always returns `None` without doing anything.
-    pub async fn from_directory_picker() -> Option<Self> {
+    pub async fn from_folder_picker() -> Result<Self> {
         if !Self::filesystem_supported() {
-            return None;
+            return Err(Error::Wasm32FilesystemNotSupported);
         }
-        let (oneshot_tx, oneshot_rx) = oneshot::channel();
+        let (oneshot_tx, oneshot_rx) = flume::bounded(1);
```

Comment:
Might be better to use actual oneshot channels since they're implemented a lot more efficiently. [async-oneshot](https://crates.io/crates/async-oneshot) seems nice. Its receivers don't have a blocking receive method, but you can do that by using `futures::executor::block_on` or by using a different, synchronous oneshot channel implementation for those cases.

Reply:
I'll try using the `oneshot` crate as it supports both async and sync channels